### PR TITLE
[FEATURE] Add support for the `dvh`, `lvh` and `svh` length units

### LINKED ISSUE #412: Add support for new viewport unit in Size class
Hi everyone,
it might be useful to add support for new viewport units (svh, lvh, svw, lvw, dvh, dvw) into the ABSOLUTE_SIZE_UNITS array in src\Value\Size.php

`const ABSOLUTE_SIZE_UNITS = ['px', 'cm', 'mm', 'mozmm', 'in', 'pt', 'pc', 'vh', 'vw', 'vmin', 'vmax', 'rem'];`

https://redstapler.co/all-css-viewport-units-svh-lvh-svw-lvw-dvh-dvw-explained/

### PR DESCRIPTION
**Pull request by KaiOnGitHub**
Fixes #412

### CODE REVIEW

File: src/Value/Size.php [hunk 1]
```diff
@@ -17,7 +17,7 @@ class Size extends PrimitiveValue
      *
      * @var array<int, string>
      */
-    const ABSOLUTE_SIZE_UNITS = ['px', 'cm', 'mm', 'mozmm', 'in', 'pt', 'pc', 'vh', 'vw', 'vmin', 'vmax', 'rem'];
+    const ABSOLUTE_SIZE_UNITS = ['px', 'cm', 'mm', 'mozmm', 'in', 'pt', 'pc', 'vh', 'dvh', 'svh', 'lvh', 'vw', 'vmin', 'vmax', 'rem'];
```
Comment (oliverklee) [inline src/Value/Size.php:20]: Could you also split split the line to below 120 character so the code sniffer is happy? Thanks!

**Comment by JakeQZ** [comment]
@KaiOnGitHub, thanks for providing the code changes and testing it within your app.  If you still don't have time to add the unit tests, I may be able to pick this up - please don't delete your branch!

**Comment by JakeQZ** [comment]
@oliverklee, I tried to rebase this branch, but couldn't figure out what I was supposed to do.  This time, I will simply re-apply the changes as a separate PR.  But for future reference, what was I doing wrong here?:

```cmd
d:\git_files\MyIntervals\PHP-CSS-Parser>git pull
remote: Enumerating objects: 1, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 1 (delta 0), reused 1 (delta 0), pack-reused 0
Unpacking objects: 100% (1/1), 521 bytes | 57.00 KiB/s, done.
From https://github.com/MyIntervals/PHP-CSS-Parser
 * [new tag]         v8.5.1     -> v8.5.1
Already up to date.

d:\git_files\MyIntervals\PHP-CSS-Parser>git fetch origin pull/415/head:pr/415
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 10 (delta 4), reused 6 (delta 4), pack-reused 0
Unpacking objects: 100% (10/10), 3.81 KiB | 69.00 KiB/s, done.
From https://github.com/MyIntervals/PHP-CSS-Parser
 * [new ref]         refs/pull/415/head -> pr/415

d:\git_files\MyIntervals\PHP-CSS-Parser>git checkout pr/415
Switched to branch 'pr/415'

d:\git_files\MyIntervals\PHP-CSS-Parser>git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

d:\git_files\MyIntervals\PHP-CSS-Parser>git pull
Already up to date.

d:\git_files\MyIntervals\PHP-CSS-Parser>git checkout pr/415
Switched to branch 'pr/415'

d:\git_files\MyIntervals\PHP-CSS-Parser>git pull
There is no tracking information for the current branch.
Please specify which branch you want to rebase against.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> pr/415


d:\git_files\MyIntervals\PHP-CSS-Parser>git branch --set-upstream-to=origin/main pr/415
branch 'pr/415' set up to track 'origin/main'.

d:\git_files\MyIntervals\PHP-CSS-Parser>git pull
Auto-merging src/Value/Size.php
CONFLICT (content): Merge conflict in src/Value/Size.php
error: could not apply 16b5f18... Add support for dvh, svh, lvh. Close #412
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 16b5f18... Add support for dvh, svh, lvh. Close #412

d:\git_files\MyIntervals\PHP-CSS-Parser>git checkout main
error: you need to resolve your current index first
src/Value/Size.php: needs merge

d:\git_files\MyIntervals\PHP-CSS-Parser>git add Size.php
fatal: pathspec 'Size.php' did not match any files

d:\git_files\MyIntervals\PHP-CSS-Parser>git add *.php

d:\git_files\MyIntervals\PHP-CSS-Parser>git rebase --continue
[detached HEAD 7039532] Add support for dvh, svh, lvh. Close #412
 Author: Kai König <50620424+KaiOnGitHub@users.noreply.github.com>
 1 file changed, 17 insertions(+), 1 deletion(-)
Auto-merging src/Value/Size.php
CONFLICT (content): Merge conflict in src/Value/Size.php
error: could not apply bc058a4... fix 120 char per line limit
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply bc058a4... fix 120 char per line limit

d:\git_files\MyIntervals\PHP-CSS-Parser>git push -f
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push origin HEAD:<name-of-remote-branch>


d:\git_files\MyIntervals\PHP-CSS-Parser>git push origin HEAD:pr/415
error: The destination you provided is not a full refname (i.e.,
starting with "refs/"). We tried to guess what you meant by:

- Looking for a ref that matches 'pr/415' on the remote side.
- Checking if the <src> being pushed ('HEAD')
  is a ref in "refs/{heads,tags}/". If so we add a corresponding
  refs/{heads,tags}/ prefix on the remote side.

Neither worked, so we gave up. You must fully qualify the ref.
hint: The <src> part of the refspec is a commit object.
hint: Did you mean to create a new branch by pushing to
hint: 'HEAD:refs/heads/pr/415'?
error: failed to push some refs to 'https://github.com/MyIntervals/PHP-CSS-Parser.git'

d:\git_files\MyIntervals\PHP-CSS-Parser>git push origin HEAD:refs/pull/415/head
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 8 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 576 bytes | 576.00 KiB/s, done.
Total 5 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/MyIntervals/PHP-CSS-Parser.git
 ! [remote rejected] HEAD -> refs/pull/415/head (deny updating a hidden ref)
error: failed to push some refs to 'https://github.com/MyIntervals/PHP-CSS-Parser.git'

d:\git_files\MyIntervals\PHP-CSS-Parser>git add *.php

d:\git_files\MyIntervals\PHP-CSS-Parser>git rebase --continue
Successfully rebased and updated refs/heads/pr/415.

d:\git_files\MyIntervals\PHP-CSS-Parser>git push -f
fatal: The upstream branch of your current branch does not match
the name of your current branch.  To push to the upstream branch
on the remote, use

    git push origin HEAD:main

To push to the branch of the same name on the remote, use

    git push origin HEAD

To choose either option permanently, see push.default in 'git help config'.

To avoid automatically configuring an upstream branch when its name
won't match the local branch, see option 'simple' of branch.autoSetupMerge
in 'git help config'.


d:\git_files\MyIntervals\PHP-CSS-Parser>git push origin HEAD
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 8 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 576 bytes | 576.00 KiB/s, done.
Total 5 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
remote:
remote: Create a pull request for 'pr/415' on GitHub by visiting:
remote:      https://github.com/MyIntervals/PHP-CSS-Parser/pull/new/pr/415
remote:
To https://github.com/MyIntervals/PHP-CSS-Parser.git
 * [new branch]      HEAD -> pr/415
```

**Comment by JakeQZ** [comment]
```md
Step 1: From your project repository, check out a new branch and test the changes.

git checkout -b KaiOnGitHub-patch-1 main
git pull https://github.com/KaiOnGitHub/PHP-CSS-Parser.git patch-1

Step 2: Merge the changes and update on GitHub.

git checkout main
git merge --no-ff KaiOnGitHub-patch-1
git push origin main
```

Trying this now.

**Comment by JakeQZ** [comment]
> ```
> Step 1: From your project repository, check out a new branch and test the changes.
> 
> git checkout -b KaiOnGitHub-patch-1 main
> git pull https://github.com/KaiOnGitHub/PHP-CSS-Parser.git patch-1
> 
> Step 2: Merge the changes and update on GitHub.
> 
> git checkout main
> git merge --no-ff KaiOnGitHub-patch-1
> git push origin main
> ```
> 
> Trying this now.

I can see this is totally stupid.  It will clearly merge the changes into `main` without review.  Fortunately I already know that the suggestions from `git` and GitHub are usually utter rubbish, so have not done this.

Much simpler to just start over.  Avoids a waste of time with systems that don't work or cannot be understood.

**Comment by oliverklee** [comment]
@JakeQZ Maybe the GitHub documentation for working on a PR branch of a fork is helpful: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork

(I have to admit that I have never tries this myself.)

**Comment by sabberworm** [comment]
> @JakeQZ Maybe the GitHub documentation for working on a PR branch of a fork is helpful

Yes, if the PR submitter has enabled “allow contributions from maintainers” setting, you’ll be able to commit to their forked branch.

However, I find that, for simple merge conflicts, the web-based UI GitHub provides is usually sufficient (I used it in this case).

**Comment by JakeQZ** [comment]
> Yes, if the PR submitter has enabled “allow contributions from maintainers” setting, you’ll be able to commit to their forked branch.

I think I maybe went wrong by [fetching the remote PR into my local repo](https://stackoverflow.com/a/30584951/2511031).  Didn't have a problem after cloning their fork into a separate directory.

File: tests/Value/SizeTest.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            )
+        );
+    }
+
+    /**
+     * @test
+     *
+     * @dataProvider provideUnit
+     */
+    public function parsesUnit(string $unit)
```
Comment (oliverklee) [inline tests/Value/SizeTest.php:45]:
```suggestion
    public function parsesUnit(string $unit): void
```

File: tests/Value/SizeTest.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            'vh', 'dvh', 'svh', 'lvh',
+            'vw', 'vmin', 'vmax', 'rem',
+            '%', 'em', 'ex', 'ch', 'fr',
+            'deg', 'grad', 'rad', 's', 'ms', 'turn', 'Hz', 'kHz',
+        ];
+
+        return \array_combine(
+            $units,
+            \array_map(
+                function (string $unit): array {
```
Comment (oliverklee) [inline tests/Value/SizeTest.php:32]: We should also use the units as array keys in the outer array: If we don't do this, PHPUnit will should something like "… with data set 12" when running the tests (which is less helpful for debugging).

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,7 @@ This project adheres to [Semantic Versioning](https://semver.org/).
 
 ### Added
 
+- Support `dvh`, `lvh` and `svh` length units (#415)
```
Comment (oliverklee) [inline CHANGELOG.md:10]:
```suggestion
- Add support for the `dvh`, `lvh` and `svh` length units (#415)
```

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: tests/Value/SizeTest.php (hunk 3 above)
Comment (JakeQZ) [reply tests/Value/SizeTest.php:32]: That's what the `array_combine` does: uses the values from `$units` also as keys.  (Saves duplicating each unit as both a key and a value.)

Comment (oliverklee) [reply tests/Value/SizeTest.php:32]: Ah, I see. Thanks for the explanation!

**Review by oliverklee — APPROVED** [review]

**Pull request merged**